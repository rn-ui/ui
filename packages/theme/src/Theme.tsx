/* tslint:disable */
import deepMerge from './tools/deepMerge';
import mergeComponentAndThemes from './tools/mergeComponentAndThemes';
import { resolveIncludes } from './tools/resolveIncludes';
import normalizeTheme from './normalizer/normalizeTheme';
import defaultTheme from './resources/default';

interface Themes {
  default: object;
  light: object;
  dark: object;
}

// @ts-ignore
const themes: Themes = {};

// Privates, ideally those should be symbols
const THEME_STYLES = '@southem.theme';

const resolveStyle = (style, baseStyle) => normalizeTheme(resolveIncludes(style, baseStyle));

/**
 * Theme
 *
 * @example
 * const defaultTheme = {};
 * const defaultThemePro = {};
 *
 * Theme.registerTheme('default', [
 *   defaultTheme,
 *   defaultThemePro,
 * ]);
 */
class Theme {

  constructor(themeStyle) {
    // @ts-ignore
    // this[THEME_STYLES] = resolveStyle(themeStyle);
  }

  /**
   * registerTheme
   * @param {String} themeName
   * @param {Array|Object} styleSheets
   */
  static registerTheme(themeName, styleSheets: Array<Object>|Object): void {
    let mergedStyleSheets = styleSheets;

    if (Array.isArray(styleSheets)) {
      mergedStyleSheets = deepMerge({}, ...styleSheets);
    }

    // themes[themeName] = new Theme(themeName, mergedStyleSheets);
    // @ts-ignore
    themes[themeName] = resolveStyle(mergedStyleSheets);
  }

  /**
   * registerDefaultTheme
   * @param {Object} styleSheets
   */
  static registerDefaultTheme(styleSheets): void {
    Theme.registerTheme('default', styleSheets);
  }

  /**
   * getTheme
   * @param {String} themeName
   * @return {Object}
   */
  static getTheme(themeName): Object {
    // @ts-ignore
    if (themeName === 'default' && !themes.default) {
      Theme.registerDefaultTheme(defaultTheme);
    }

    return themes[themeName];
  }

  /**
   * getDefaultTheme
   * @return {Object}
   */
  static getDefaultTheme(): Object {
    return Theme.getTheme('default');
  }

  /**
   * isDefine
   * @param {String} themeName
   * @return {Boolean}
   */
  static isDefine(themeName): Boolean {
    return !!themes[themeName];
  }
}

export {
  Theme as default,
  themes,
};
