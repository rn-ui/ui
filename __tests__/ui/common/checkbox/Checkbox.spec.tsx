import React from 'react';
import {
  Text,
  TouchableOpacity,
} from 'react-native';
import {
  fireEvent,
  render,
  RenderAPI,
  shallow,
} from 'react-native-testing-library';
import Theme, {
  ThemeProvider,
  ThemeProviderProps,
} from '@southem/theme';
import {
  darkTheme,
  CheckBox,
  CheckBoxProps,
} from '@southem/ui';


Theme.registerDefaultTheme(darkTheme);

const CheckBoxMock = (props?: CheckBoxProps): React.ReactElement<ThemeProviderProps> => {
  return (
    <ThemeProvider
      theme={'default'}>
      <CheckBox {...props} />
    </ThemeProvider>
  );
};

const renderComponent = (props?: CheckBoxProps): RenderAPI => {
  return render(
    <CheckBoxMock {...props} />,
  );
};

describe('@checkbox: component checks', () => {
  it('* should request checking', () => {
    const onCheckedChange = jest.fn();
    const component = renderComponent({
      checked: false,
      onChange: onCheckedChange,
    });

    fireEvent.press(component.queryByType(TouchableOpacity));
    expect(onCheckedChange).toBeCalledWith(true, false);
  });

  it('* should request unchecking', () => {
    const onCheckedChange = jest.fn();
    const component = renderComponent({
      checked: true,
      onChange: onCheckedChange,
      });

    fireEvent.press(component.queryByType(TouchableOpacity));
    expect(onCheckedChange).toBeCalledWith(false, false);
  });

  it('* should request clearing indeterminate and checking', () => {
    const onCheckedChange = jest.fn();
    const component = renderComponent({
      checked: false,
      indeterminate: true,
      onChange: onCheckedChange,
    });

    fireEvent.press(component.queryByType(TouchableOpacity));
    expect(onCheckedChange).toBeCalledWith(true, false);
  });

  it('* should request clearing indeterminate and unchecking', () => {
    const onCheckedChange = jest.fn();
    const component = renderComponent({
      checked: true,
      indeterminate: true,
      onChange: onCheckedChange,
    });

    fireEvent.press(component.queryByType(TouchableOpacity));
    expect(onCheckedChange).toBeCalledWith(false, false);
  });

  it('* should render text', () => {
    const component = renderComponent({
      checked: false,
      children: <Text>I love Babel</Text>,
    });
    const { output } = shallow(component.getByText('I love Babel'));

    expect(output).toBeTruthy();
    expect(output).toMatchSnapshot();
  });

  it('* should render text as component', () => {
    const component = render(
      <CheckBoxMock checked>
        {props => <Text {...props}>I love Babel</Text>}
      </CheckBoxMock>,
    );

    const { output } = shallow(component.getByText('I love Babel'));

    expect(output).toBeTruthy();
    expect(output).toMatchSnapshot();
  });

  it('* should call onPressIn', () => {
    const onPressIn = jest.fn();
    const component = renderComponent({
      onPressIn,
    });

    fireEvent(component.queryByType(TouchableOpacity), 'pressIn');
    expect(onPressIn).toBeCalled();
  });

  it('* should call onPressOut', () => {
    const onPressOut = jest.fn();
    const component = renderComponent({
      onPressOut,
    });

    fireEvent(component.queryByType(TouchableOpacity), 'pressOut');
    expect(onPressOut).toBeCalled();
  });

  it('* should call onFocus', () => {
    const onFocus = jest.fn();

    const component = renderComponent({
      onFocus,
    });

    fireEvent(component.queryByType(TouchableOpacity), 'focus');
    expect(onFocus).toBeCalled();
  });

  it('* should call onBlur', () => {
    const onBlur = jest.fn();

    const component = renderComponent({
      onBlur,
    });

    fireEvent(component.queryByType(TouchableOpacity), 'blur');
    expect(onBlur).toBeCalled();
  });

});
