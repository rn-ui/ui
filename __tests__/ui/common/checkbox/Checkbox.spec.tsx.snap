// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@checkbox: component checks * should render text 1`] = `
<Text
  allowFontScaling={false}
  animationOptions={Object {}}
  containerStyle={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
  onLayout={[Function]}
  style={
    Array [
      Object {
        "color": "rgba(0, 0, 0, 0.5)",
        "fontSize": 24,
      },
      Object {},
      Object {
        "fontFamily": "Material Icons",
        "fontStyle": "normal",
        "fontWeight": "normal",
      },
      Object {},
    ]
  }
>
  ?
</Text>
`;

exports[`@checkbox: component checks * should render text as component 1`] = `
<Text
  allowFontScaling={false}
  animationOptions={Object {}}
  containerStyle={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
  onLayout={[Function]}
  style={
    Array [
      Object {
        "color": "rgba(0, 0, 0, 0.5)",
        "fontSize": 24,
      },
      Object {},
      Object {
        "fontFamily": "Material Icons",
        "fontStyle": "normal",
        "fontWeight": "normal",
      },
      Object {},
    ]
  }
>
  ?
</Text>
`;
